### ISSUE TITLE
Question: Plotting of xarray.Dataset with multiple data variables (multiple vdims)

### ISSUE THREAD

**Issue by SebastianRiedel** [ioam/holoviews#2015]
The natural format of my data looks something like this. Two observed variables, tau and g, over time and per joint:
```
<xarray.Dataset>
Dimensions:  (joint: 7, time: 100)
Coordinates:
  * joint    (joint) int64 0 1 2 3 4 5 6
  * time     (time) int64 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 ...
Data variables:
    tau      (time, joint) float64 0.7896 0.8261 0.4972 0.639 0.4531 0.2167 ...
    g        (time, joint) float64 0.6047 0.5693 1.25 0.8547 0.5194 0.9736 ...
```
**The plot I want is tau and g overlayed (over time) and a slider to select over the joint dimension.** This makes sense as in my case tau and g have the same physical unit (Nm) and one is the predicted and the other the measured torque value of a physical system.

The easiest way to do this, however, seems to be to bring the dataset into an artificial single-variable form, introducing an additional dimension telemetry_type:
```
<xarray.Dataset>
Dimensions:          (joint: 7, telemetry_type: 2, time: 100)
Coordinates:
  * joint            (joint) int64 0 1 2 3 4 5 6
  * time             (time) int64 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 ...
  * telemetry_type   (telemetry_type) object 'tau' 'g'
Data variables:
    telemetry_value  (telemetry_type, time, joint) float64 0.7896 0.8261 ...
```
which allows to plot it comfortably with
```
hv.Dataset(single_var_dataset).to(hv.Curve, kdims=['time']).overlay('telemetry_type')
```
![9563875c-5e2e-4efc-af24-7c881bdbbaba](https://user-images.githubusercontent.com/3171935/31866711-f89faca0-b783-11e7-86b6-39c605500ee3.png)

Is there an easy way to do the same kind of plot with the original multi-variable xarray.Dataset? I tried something like below (which produces an identical plot to above), but it seems not particularly elegant. What am I missing? How can I use the a hv.Dataset object to overlay value dimensions, not key dimensions?
```
def vdims_overlay(joint):
    return hv.NdOverlay({dim: hv.Curve((dset.sel(joint=joint)['time'], dset.sel(joint=joint)[dim]), kdims=['time'], vdims=['value']) for dim in dset.data_vars.keys()})

dm = hv.DynamicMap(vdims_overlay, kdims=['joint'])
dm.redim.values(joint=dset['joint'].values)
```

**Comment by philippjfr** [comment]
This is a very general problem that also frequently applies to pandas DataFrames, so thanks for filing this. I'll try to put together a proposal for a ``Dataset.melt`` method that will let you go from a "wide" dataset like yours to a "tidy" one. Conceptually using ``tidy`` datasets is the correct way to use HoloViews but there absolutely needs to be an easy way to make a dataset conform to that expectation.

**Comment by philippjfr** [comment]
If you don't mind I'll repurpose this issue to one more generally about working with "wide" datasets.

**Comment by SebastianRiedel** [comment]
I might be wrong, but I'm not sure if this is an issue with the dataset representation alone. I know what you mean by a [tidy dataset](http://vita.had.co.nz/papers/tidy-data.html).

It can make sense to melt such a dataset (e.g. with the frequency of a disease as in the [Tabular Datasets Tutorial](http://holoviews.org/getting_started/Tabular_Datasets.html) or in the tidy dataset paper, Table 1). But if I store quantities with different physical units or if these are indeed measurements of different variables (like my example above), it does not seem logical to me to melt those columns. See also the tidy dataset paper, Table 12 b), for an example of this kind of tidy data. 

This might be a more general issue with plotting datasets with multiple vdims? Would it make sense that HoloMap-layout methods like ```.overlay, .grid, .layout``` can work with not only key, but value dimensions, e.g. by specifying a list of vdims? This would allow something like:
```
hv.Dataset(multi_vdims_dataset).to(hv.Curve, kdims=['time']).overlay(vdims=['tau', 'g'])
hv.Dataset(multi_vdims_dataset).to(hv.Curve, kdims=['time']).overlay(vdims=['*']) # or a different nice shortcut for all vdims
```

**Comment by philippjfr** [comment]
>This might be a more general issue with plotting datasets with multiple vdims?

Yes, it's an issue that comes up fairly frequently and we should get this right.

>Would it make sense that HoloMap-layout methods like .overlay, .grid, .layout can work with not only key, but value dimensions, e.g. by specifying a list of vdims? This would allow something like:

Sure, I agree that might be a good approach.

>But if I store quantities with different physical units or if these are indeed measurements of different variables (like my example above), it does not seem logical to me to melt those columns.

I wonder about this a little bit though, if the variables really have different physical units does it really make sense to overlay them (ignoring the introduction of twin axes for now)? In general HoloViews is designed with the assumption that a dimensioned container like a HoloMap, NdLayout or NdOverlay contains Elements with matching dimensions. While this is not strictly enforced it does mean that you might get non-intuitive behavior if you break that assumption.

I strongly agree that we need to do something to make this easier. I'm also open to relaxing these assumptions and making it easier to word with "wide" data in the way you describe, I just want to be sure it's definitely the right approach.

**Comment by SebastianRiedel** [comment]
You're right. Different physical units was just meant as a motivation for datasets with multiple vdims. Overlaying won't make sense in this case in general (ignoring twin axis :)). I'm not sure at the moment either, where multiple vdims should be handled best.

**Comment by nbren12** [comment]
I would just add that I often need to overlay/plot side-by-side  data with different physical units. This often happens two dimensional maps. For instance, I could produce an image plot of sea level pressure (Pa) and the overlay contours of surface temperature (K), or I could plot them side by side.
